mirrord/intproxy/src/proxies/incoming/http.rs

```diff
@@ -1,53 +1,89 @@
-use std::convert::Infallible;
+use std::{convert::Infallible, future};
 
 use bytes::Bytes;
 use http_body_util::combinators::BoxBody;
 use hyper::{
     body::Incoming,
-    client::conn::{http1, http2},
+    client::conn::{
+        http1::{self, Parts},
+        http2,
+    },
     Response, Version,
 };
 use hyper_util::rt::{TokioExecutor, TokioIo};
 use mirrord_protocol::tcp::HttpRequestFallback;
-use tokio::net::TcpStream;
+use tokio::{
+    net::TcpStream,
+    sync::oneshot::{self, Receiver},
+};
+
+use super::interceptor::{Error, Result};
 
-use super::http_interceptor::HttpInterceptorError;
+/// HTTP connection deconstructed after an `UPGRADE`.
+pub struct UpgradedConnection {
+    /// Raw TCP stream that was used to send HTTP requests.
+    pub stream: TcpStream,
+    /// Bytes that were read by [`hyper`] from the server, but were not processed as HTTP.
+    pub unprocessed_bytes: Bytes,
+}
 
 /// Handles the differences between hyper's HTTP/1 and HTTP/2 connections.
-pub enum HttpConnection {
+pub enum HttpSender {
     V1(http1::SendRequest<BoxBody<Bytes, Infallible>>),
     V2(http2::SendRequest<BoxBody<Bytes, Infallible>>),
 }
 
-impl HttpConnection {
-    pub async fn handshake(
-        version: Version,
-        target_stream: TcpStream,
-    ) -> Result<Self, HttpInterceptorError> {
-        match version {
-            Version::HTTP_2 => {
-                let (sender, connection) =
-                    http2::handshake(TokioExecutor::default(), TokioIo::new(target_stream)).await?;
-                tokio::spawn(connection);
+/// Consumes the given [`TcpStream`] and performs an HTTP handshake, turning it into an HTTP
+/// connection.
+///
+/// # Returns
+///
+/// * [`HttpSender`] that can be used to send HTTP requests to the peer.
+/// * [`Receiver`] that can be used to reclaim the [`TcpStream`] after the HTTP connection ends with
+///   an `UPGRADE`.
+pub async fn handshake(
+    version: Version,
+    target_stream: TcpStream,
+) -> Result<(HttpSender, Receiver<Result<UpgradedConnection>>)> {
+    match version {
+        Version::HTTP_2 => {
+            let (sender, connection) =
+                http2::handshake(TokioExecutor::default(), TokioIo::new(target_stream)).await?;
+            tokio::spawn(connection);
 
-                Ok(Self::V2(sender))
-            }
+            let (upgrade_tx, upgrade_rx) = oneshot::channel();
+            let _ = upgrade_tx.send(Err(Error::UpgradeNotSupported(version)));
```

Comment:
I know that this might be a sad request, but feels a bit weird to preemptively send a "can't upgrade" here like this, and I'm assuming whoever takes the `upgrade_rx` deals with it when calling `recv`?

Why not express this as an `Option<Receiver>` in the return of `handshake`?